src/lib.rs

```diff
@@ -1,7 +1,1234 @@
+    fn path() {
+        let seed = [0; 32];
+        let xsk_m = ExtendedSpendingKey::master(&seed);
+
+        let xsk_5h = xsk_m.derive_child(ChildIndex::Hardened(5));
+        assert_eq!(
+            ExtendedSpendingKey::from_path(&xsk_m, &[ChildIndex::Hardened(5)]),
+            xsk_5h
+        );
+
+        let xsk_5h_7 = xsk_5h.derive_child(ChildIndex::NonHardened(7));
+        assert_eq!(
+            ExtendedSpendingKey::from_path(
+                &xsk_m,
+                &[ChildIndex::Hardened(5), ChildIndex::NonHardened(7)]
+            ),
+            xsk_5h_7
+        );
+    }
+
+    #[test]
+    fn diversifier() {
+        let dk = DiversifierKey([0; 32]);
+        let j_0 = DiversifierIndex::new();
+        let j_1 = DiversifierIndex([1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]);
+        let j_2 = DiversifierIndex([2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]);
+        let j_3 = DiversifierIndex([3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]);
+        // Computed using this Rust implementation
+        let d_0 = [220, 231, 126, 188, 236, 10, 38, 175, 214, 153, 140];
+        let d_3 = [60, 253, 170, 8, 171, 147, 220, 31, 3, 144, 34];
+
+        // j = 0
+        let (j, d_j) = dk.diversifier(j_0).unwrap();
+        assert_eq!(j, j_0);
+        assert_eq!(d_j.0, d_0);
+
+        // j = 1
+        let (j, d_j) = dk.diversifier(j_1).unwrap();
+        assert_eq!(j, j_3);
+        assert_eq!(d_j.0, d_3);
+
+        // j = 2
+        let (j, d_j) = dk.diversifier(j_2).unwrap();
+        assert_eq!(j, j_3);
+        assert_eq!(d_j.0, d_3);
+
+        // j = 3
+        let (j, d_j) = dk.diversifier(j_3).unwrap();
+        assert_eq!(j, j_3);
+        assert_eq!(d_j.0, d_3);
+    }
+
+    #[test]
+    fn default_address() {
+        let seed = [0; 32];
+        let xsk_m = ExtendedSpendingKey::master(&seed);
+        let (j_m, addr_m) = xsk_m.default_address().unwrap();
+        assert_eq!(j_m.0, [0; 11]);
+        assert_eq!(
+            addr_m.diversifier.0,
+            // Computed using this Rust implementation
+            [59, 246, 250, 31, 131, 191, 69, 99, 200, 167, 19]
+        );
+    }
+
+    #[test]
+    fn read_write() {
+        let seed = [0; 32];
+        let xsk = ExtendedSpendingKey::master(&seed);
+        let fvk = ExtendedFullViewingKey::from(&xsk);
+
+        let mut ser = vec![];
+        xsk.write(&mut ser).unwrap();
+        let xsk2 = ExtendedSpendingKey::read(&ser[..]).unwrap();
+        assert_eq!(xsk2, xsk);
+
+        let mut ser = vec![];
+        fvk.write(&mut ser).unwrap();
+        let fvk2 = ExtendedFullViewingKey::read(&ser[..]).unwrap();
+        assert_eq!(fvk2, fvk);
+    }
+
     #[test]
-    fn it_works() {
-        assert_eq!(2 + 2, 4);
+    fn test_vectors() {
+        struct TestVector {
+            ask: Option<[u8; 32]>,
+            nsk: Option<[u8; 32]>,
+            ovk: [u8; 32],
+            dk: [u8; 32],
+            c: [u8; 32],
+            ak: [u8; 32],
+            nk: [u8; 32],
+            ivk: [u8; 32],
```

Comment:
The compiler complains about the following:
```
warning: field is never used: `ivk`
   --> src\lib.rs:738:13
```

Reply:
I forgot to add a test checking `ivk`. Added.